### PR TITLE
Media: add FIFO buffer for audio input stream

### PR DESCRIPTION
**Pull request by sideb0ard**
Add FIFO buffer to handle case when input stream is bursty.

Tested on Android and Linux implementations.

Bug: 483713292

### CODE REVIEW

File: media/webrtc/audio_processor.cc [hunk 1]
```diff
@@ -620,6 +620,8 @@ AudioParameters AudioProcessor::GetDefaultOutputFormat(
 #if BUILDFLAG(IS_CASTOS) || BUILDFLAG(IS_CAST_ANDROID)
           std::min(media::WebRtcAudioProcessingSampleRateHz(),
                    input_format.sample_rate())
+#elif BUILDFLAG(IS_COBALT)
+          input_format.sample_rate()
```
Comment (kjyoun) [inline media/webrtc/audio_processor.cc]: I think that this is a separate change than changes in audio_input_stream_starboard.h

Can you make this as a separate PR?

File: base/sync_socket_posix.cc [hunk 2]
```diff
@@ -174,7 +174,9 @@ size_t SyncSocket::Peek() {
   ssize_t number_chars = recv(handle_.get(), buffer, sizeof(buffer),
                               MSG_PEEK | MSG_TRUNC | MSG_DONTWAIT);
   if (number_chars < 0) {
-    PLOG(ERROR) << "recv failed in SyncSocket::Peek";
+    if (errno != EAGAIN && errno != EWOULDBLOCK) {
+      PLOG(ERROR) << "recv failed in SyncSocket::Peek";
+    }
```
Comment (kjyoun) [inline base/sync_socket_posix.cc]: Can you make this change into separate PR?

That being said, how can we verify this is a right change?

why not try again when EAGAIN or EWOULDBLOCK?

Can you add comment about this behavior?

another question --
why do we need starboard own impl.. not using chromium's impl. below (iotcl)?

File: media/audio/starboard/audio_input_stream_starboard.cc [hunk 3]
```diff
@@ -14,12 +14,18 @@
...
+#include <algorithm>
+
 #include "base/logging.h"
 #include "media/audio/audio_manager.h"
 #include "media/audio/audio_manager_base.h"
 
 namespace media {
 
+namespace {
+constexpr int kMinReadSizeBytes = 2048;
```
Comment (kjyoun) [inline media/audio/starboard/audio_input_stream_starboard.cc]: super nit: `kMinReadBytes` ?

File: media/audio/starboard/audio_input_stream_starboard.cc [hunk 4]
```diff
@@ -171,65 +176,62 @@ void AudioInputStreamStarboard::ReadAudio() {
...
-                       base::Unretained(this)),
-        next_check_time, base::subtle::DelayPolicy::kPrecise);
-  } else {
-    DLOG(WARNING) << "SbMicrophoneRead returned " << bytes_read
-                  << ". Dropping this buffer.";
+  } else if (bytes_read < 0) {
+    DLOG(WARNING) << "SbMicrophoneRead returned " << bytes_read << ".";
     HandleError("SbMicrophoneRead");
     return;
   }
```
Comment (kjyoun) [inline media/audio/starboard/audio_input_stream_starboard.cc:209]: don't we need to handle bytes_read == 0?

**Review by kjyoun — COMMENTED** [review]
I think that this PR has 3 different changes, each of which deserves its own PR and review.

Can you break down?

super nit: can u change title to "media: Add... " per https://g3doc.corp.google.com/company/teams/cobalt/team/coding_style.md?cl=head#commit-message

File: base/sync_socket_posix.cc (hunk 2 above)
Comment (sideb0ard) [reply base/sync_socket_posix.cc]: new PR at https://github.com/youtube/cobalt/pull/10470

> why do we need starboard own impl.. not using chromium's impl. below (iotcl)?

Starboard doesn't support ioctl
https://github.com/youtube/cobalt/pull/6641/changes

File: media/webrtc/audio_processor.cc (hunk 1 above)
Comment (sideb0ard) [reply media/webrtc/audio_processor.cc]: done - https://github.com/youtube/cobalt/pull/10471

File: media/audio/starboard/audio_input_stream_starboard.cc (hunk 4 above)
Comment (sideb0ard) [reply media/audio/starboard/audio_input_stream_starboard.cc:209]: We don't need an explicit  else if (bytes_read == 0)  branch anymore because it is handled correctly by the fall-through behavior of the new code.
  
Here is what happens when  bytes_read  is  0 :
  
1. It skips  if (bytes_read > 0)  (no data to add to FIFO).
2. It skips  else if (bytes_read < 0)  (no error).
3. It checks if the FIFO has enough data to push ( audio_fifo_.size() >= target_buffer_size_bytes ).
      • If it does (from previous reads), it pushes it and sets  data_pushed = true .
      • If it doesn't,  data_pushed  remains  false .
4. Finally, it hits the adaptive scheduling:
      • Since no data was read and (presumably) no data was pushed,  data_pushed  is false, and it schedules the next read in  buffer_duration_ / 2 .
  
This matches the old behavior for  bytes_read == 0  (polling faster to check for data again) but with the added benefit that we can still push a chunk if there was enough data left over in the FIFO from a previous burst.

File: media/audio/starboard/audio_input_stream_starboard.cc (hunk 3 above)
Comment (sideb0ard) [reply media/audio/starboard/audio_input_stream_starboard.cc]: updated.

File: media/audio/starboard/audio_input_stream_starboard.h [hunk 5]
```diff
@@ -74,7 +76,7 @@ class AudioInputStreamStarboard : public AgcAudioStream<AudioInputStream> {
   bool closing_ = false;
   std::unique_ptr<AudioBus> audio_bus_;
   std::vector<int16_t> buffer_;
-  int frames_in_buffer_ = 0;
+  std::vector<uint8_t> audio_fifo_;
```
[the remark this answers is not in the document]
Comment (sideb0ard) [reply media/audio/starboard/audio_input_stream_starboard.h:79]: will stick with vector

File: media/audio/starboard/audio_input_stream_starboard.cc [hunk 6]
```diff
@@ -171,65 +176,62 @@ void AudioInputStreamStarboard::ReadAudio() {
...
+        std::max(base::TimeTicks::Now(), next_read_time_ + buffer_duration_);
+  } else {
+    next_read_time_ = base::TimeTicks::Now() + buffer_duration_ / 2;
+  }
+
+  capture_thread_.task_runner()->PostDelayedTaskAt(
+      base::subtle::PostDelayedTaskPassKey(), FROM_HERE,
+      base::BindOnce(&AudioInputStreamStarboard::ReadAudio,
+                     base::Unretained(this)),
+      next_read_time_, base::subtle::DelayPolicy::kPrecise);
```
[the remark this answers is not in the document]
Comment (sideb0ard) [reply media/audio/starboard/audio_input_stream_starboard.cc:237]: refactored

File: media/audio/starboard/audio_input_stream_starboard.cc [hunk 7]
```diff
@@ -171,65 +176,65 @@ void AudioInputStreamStarboard::ReadAudio() {
...
+    }
+  }
+
+  bool data_pushed = false;
+  // Process all available complete buffers to minimize latency.
+  while (audio_fifo_.size() >= target_buffer_size_bytes) {
+    std::copy(audio_fifo_.begin(),
+              audio_fifo_.begin() + target_buffer_size_bytes,
+              reinterpret_cast<uint8_t*>(buffer_.data()));
+    audio_fifo_.erase(audio_fifo_.begin(),
```
Comment (kjyoun) [inline media/audio/starboard/audio_input_stream_starboard.cc:217]: std::vector::erase(..) has overhead of O(n). Inside of the loop, it results in O(N^2)


How about using `base::circular_deque` which can handle erase O(1) ?

File: media/audio/starboard/audio_input_stream_starboard.cc [hunk 8]
```diff
@@ -171,65 +176,65 @@ void AudioInputStreamStarboard::ReadAudio() {
...
-      frames_to_read * params_.channels() * sizeof(int16_t);
+  const size_t target_buffer_size_bytes =
+      params_.frames_per_buffer() * params_.channels() * sizeof(int16_t);
+  const size_t max_fifo_size_bytes = target_buffer_size_bytes * kMaxFifoBuffers;
+  const size_t read_size_bytes =
+      std::max(static_cast<size_t>(kMinReadBytes), target_buffer_size_bytes);
 
+  // Read directly into the FIFO to avoid extra allocation and copy.
+  size_t current_fifo_size = audio_fifo_.size();
+  audio_fifo_.resize(current_fifo_size + read_size_bytes);
```
Comment (kjyoun) [inline media/audio/starboard/audio_input_stream_starboard.cc:187]: I can see  you are calling resize() in several places. here and line 194, and line 203

Can we resize only once to the possible largest size?

calling resize() multiple times per each Read() does not look ideal.

File: media/audio/starboard/audio_input_stream_starboard.cc
Comment (kjyoun) [inline media/audio/starboard/audio_input_stream_starboard.cc:172]: The FIFO buffering logic here is getting quite complex and handles some tricky edge cases (like overflow and partial buffer accumulation). This logic deserves its own isolated unit tests.
 
How about extract this into a dedicated  class (e..g. AudioFifo) and cover it with unit tests, and have  AudioInputStreamStarboard simply use that class? 

It will  improve the readability of  ReadAudio()  and ensure our buffer math is bulletproof
